File: y_eet-synth/synth/metrics.py
"""
In-process metrics collection.
Thread/coroutine-safe counters and latency histograms that produce
p50/p95/p99 without external dependencies.
"""

from __future__ import annotations

import asyncio
import statistics
import time
from collections import defaultdict
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class RequestRecord:
    endpoint: str
    method: str
    status_code: int
    latency_ms: float
    retry_count: int = 0  # x-envoy-attempt-count - 1
    idempotency_replay: bool = False
    timeout: bool = False
    auth_failed: bool = False
    trace_id: str = ""
    traceparent_sent: bool = False
    traceparent_received: bool = False
    canary_version: str = ""  # value of x-canary-version (or equivalent)
    envoy_upstream_ms: Optional[float] = None  # x-envoy-upstream-service-time
    envoy_attempt_count: int = 1  # x-envoy-attempt-count
    via_istio: bool = False  # server: istio-envoy


@dataclass
class LatencyStats:
    samples: list[float] = field(default_factory=list)

    def record(self, ms: float) -> None:
        self.samples.append(ms)

    def percentile(self, pct: float) -> float:
        if not self.samples:
            return 0.0
        sorted_samples = sorted(self.samples)
        idx = int(len(sorted_samples) * pct / 100)
        idx = min(idx, len(sorted_samples) - 1)
        return sorted_samples[idx]

    @property
    def p50(self) -> float:
        return self.percentile(50)

    @property
    def p95(self) -> float:
        return self.percentile(95)

    @property
    def p99(self) -> float:
        return self.percentile(99)

    @property
    def mean(self) -> float:
        return statistics.mean(self.samples) if self.samples else 0.0

    @property
    def count(self) -> int:
        return len(self.samples)


class EndpointMetrics:
    def __init__(self, endpoint: str) -> None:
        self.endpoint = endpoint
        self.latency = LatencyStats()
        self.total: int = 0
        self.success: int = 0  # 2xx
        self.client_error: int = 0  # 4xx
        self.server_error: int = 0  # 5xx
        self.timeout: int = 0
        self.retried: int = 0  # requests where attempt_count > 1
        self.retry_total: int = 0  # sum of extra attempts
        self.idempotency_hits: int = 0
        self.auth_failures: int = 0
        self.status_codes: dict[int, int] = defaultdict(int)
        self.canary_hits: dict[str, int] = defaultdict(int)
        self.trace_sent: int = 0
        self.trace_received: int = 0
        self.via_istio: int = 0

    def record(self, r: RequestRecord) -> None:
        self.total += 1
        self.latency.record(r.latency_ms)
        self.status_codes[r.status_code] += 1

        if 200 <= r.status_code < 300:
            self.success += 1
        elif 400 <= r.status_code < 500:
            self.client_error += 1
            if r.auth_failed:
                self.auth_failures += 1
        elif r.status_code >= 500:
            self.server_error += 1

        if r.timeout:
            self.timeout += 1
        if r.idempotency_replay:
            self.idempotency_hits += 1
        if r.envoy_attempt_count > 1:
            self.retried += 1
            self.retry_total += r.envoy_attempt_count - 1
        if r.canary_version:
            self.canary_hits[r.canary_version] += 1
        if r.traceparent_sent:
            self.trace_sent += 1
        if r.traceparent_received:
            self.trace_received += 1
        if r.via_istio:
            self.via_istio += 1

    @property
    def error_rate(self) -> float:
        if self.total == 0:
            return 0.0
        return (self.client_error + self.server_error + self.timeout) / self.total

    @property
    def success_rate(self) -> float:
        if self.total == 0:
            return 0.0
        return self.success / self.total

    @property
    def timeout_rate(self) -> float:
        if self.total == 0:
            return 0.0
        return self.timeout / self.total

    @property
    def avg_attempt_count(self) -> float:
        if self.total == 0:
            return 1.0
        return 1.0 + (self.retry_total / self.total)

    @property
    def trace_propagation_rate(self) -> float:
        if self.trace_sent == 0:
            return 0.0
        return self.trace_received / self.trace_sent


class MetricsCollector:
    """Global in-process metrics store. All methods are asyncio-safe."""

    def __init__(self) -> None:
        self._lock = asyncio.Lock()
        self._endpoints: dict[str, EndpointMetrics] = {}
        self._start_time: float = time.monotonic()
        self._total_records: int = 0

    async def record(self, r: RequestRecord) -> None:
        async with self._lock:
            if r.endpoint not in self._endpoints:
                self._endpoints[r.endpoint] = EndpointMetrics(r.endpoint)
            self._endpoints[r.endpoint].record(r)
            self._total_records += 1

    @property
    def elapsed_seconds(self) -> float:
        return time.monotonic() - self._start_time

    @property
    def rps(self) -> float:
        elapsed = self.elapsed_seconds
        return self._total_records / elapsed if elapsed > 0 else 0.0

    def snapshot(self) -> dict[str, EndpointMetrics]:
        return dict(self._endpoints)

    def total(self) -> int:
        return self._total_records

    def global_error_rate(self) -> float:
        snap = self.snapshot()
        total = sum(m.total for m in snap.values())
        errors = sum(m.client_error + m.server_error + m.timeout for m in snap.values())
        return errors / total if total > 0 else 0.0

    def global_p99(self) -> float:
        all_samples: list[float] = []
        for m in self.snapshot().values():
            all_samples.extend(m.latency.samples)
        if not all_samples:
            return 0.0
        s = sorted(all_samples)
        return s[int(len(s) * 0.99)]

File: y_eet-synth/internal/metrics/metrics.go
// Package metrics provides in-process request metrics collection.
package metrics

import (
	"math"
	"sort"
	"sync"
	"time"
)

// RequestRecord captures the result of a single HTTP request.
type RequestRecord struct {
	Endpoint            string
	Method              string
	StatusCode          int
	LatencyMs           float64
	RetryCount          int    // x-envoy-attempt-count - 1
	IdempotencyReplay   bool
	Timeout             bool
	AuthFailed          bool
	TraceID             string
	TraceparentSent     bool
	TraceparentReceived bool
	CanaryVersion       string
	EnvoyUpstreamMs     float64
	EnvoyAttemptCount   int
	ViaIstio            bool
}

// LatencyStats tracks latency samples and computes percentiles.
type LatencyStats struct {
	samples []float64
}

func (l *LatencyStats) Record(ms float64) {
	l.samples = append(l.samples, ms)
}

func (l *LatencyStats) Percentile(pct float64) float64 {
	if len(l.samples) == 0 {
		return 0
	}
	sorted := make([]float64, len(l.samples))
	copy(sorted, l.samples)
	sort.Float64s(sorted)
	idx := int(math.Floor(float64(len(sorted)) * pct / 100.0))
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	return sorted[idx]
}

func (l *LatencyStats) P50() float64 { return l.Percentile(50) }
func (l *LatencyStats) P95() float64 { return l.Percentile(95) }
func (l *LatencyStats) P99() float64 { return l.Percentile(99) }

func (l *LatencyStats) Mean() float64 {
	if len(l.samples) == 0 {
		return 0
	}
	sum := 0.0
	for _, s := range l.samples {
		sum += s
	}
	return sum / float64(len(l.samples))
}

func (l *LatencyStats) Count() int { return len(l.samples) }

// EndpointMetrics tracks all metrics for a single endpoint.
type EndpointMetrics struct {
	Endpoint        string
	Latency         LatencyStats
	Total           int
	Success         int // 2xx
	ClientError     int // 4xx
	ServerError     int // 5xx
	Timeout         int
	Retried         int // requests where attempt_count > 1
	RetryTotal      int // sum of extra attempts
	IdempotencyHits int
	AuthFailures    int
	StatusCodes     map[int]int
	CanaryHits      map[string]int
	TraceSent       int
	TraceReceived   int
	ViaIstio        int
}

func newEndpointMetrics(endpoint string) *EndpointMetrics {
	return &EndpointMetrics{
		Endpoint:    endpoint,
		StatusCodes: make(map[int]int),
		CanaryHits:  make(map[string]int),
	}
}

func (e *EndpointMetrics) Record(r RequestRecord) {
	e.Total++
	e.Latency.Record(r.LatencyMs)
	e.StatusCodes[r.StatusCode]++

	switch {
	case r.StatusCode >= 200 && r.StatusCode < 300:
		e.Success++
	case r.StatusCode >= 400 && r.StatusCode < 500:
		e.ClientError++
		if r.AuthFailed {
			e.AuthFailures++
		}
	case r.StatusCode >= 500:
		e.ServerError++
	}

	if r.Timeout {
		e.Timeout++
	}
	if r.IdempotencyReplay {
		e.IdempotencyHits++
	}
	if r.EnvoyAttemptCount > 1 {
		e.Retried++
		e.RetryTotal += r.EnvoyAttemptCount - 1
	}
	if r.CanaryVersion != "" {
		e.CanaryHits[r.CanaryVersion]++
	}
	if r.TraceparentSent {
		e.TraceSent++
	}
	if r.TraceparentReceived {
		e.TraceReceived++
	}
	if r.ViaIstio {
		e.ViaIstio++
	}
}

func (e *EndpointMetrics) ErrorRate() float64 {
	if e.Total == 0 {
		return 0
	}
	return float64(e.ClientError+e.ServerError+e.Timeout) / float64(e.Total)
}

func (e *EndpointMetrics) SuccessRate() float64 {
	if e.Total == 0 {
		return 0
	}
	return float64(e.Success) / float64(e.Total)
}

func (e *EndpointMetrics) TimeoutRate() float64 {
	if e.Total == 0 {
		return 0
	}
	return float64(e.Timeout) / float64(e.Total)
}

func (e *EndpointMetrics) AvgAttemptCount() float64 {
	if e.Total == 0 {
		return 1.0
	}
	return 1.0 + float64(e.RetryTotal)/float64(e.Total)
}

func (e *EndpointMetrics) TracePropagationRate() float64 {
	if e.TraceSent == 0 {
		return 0
	}
	return float64(e.TraceReceived) / float64(e.TraceSent)
}

// Collector is the global in-process metrics store.
type Collector struct {
	mu           sync.Mutex
	endpoints    map[string]*EndpointMetrics
	startTime    time.Time
	totalRecords int
}

// New creates a Collector.
func New() *Collector {
	return &Collector{
		endpoints: make(map[string]*EndpointMetrics),
		startTime: time.Now(),
	}
}

// Record records a single request result.
func (c *Collector) Record(r RequestRecord) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if _, ok := c.endpoints[r.Endpoint]; !ok {
		c.endpoints[r.Endpoint] = newEndpointMetrics(r.Endpoint)
	}
	c.endpoints[r.Endpoint].Record(r)
	c.totalRecords++
}

// ElapsedSeconds returns seconds since collection started.
func (c *Collector) ElapsedSeconds() float64 {
	return time.Since(c.startTime).Seconds()
}

// RPS returns average requests per second since collection started.
func (c *Collector) RPS() float64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	elapsed := c.ElapsedSeconds()
	if elapsed <= 0 {
		return 0
	}
	return float64(c.totalRecords) / elapsed
}

// Total returns total requests recorded.
func (c *Collector) Total() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.totalRecords
}

// Snapshot returns a copy of the endpoint metrics map.
func (c *Collector) Snapshot() map[string]*EndpointMetrics {
	c.mu.Lock()
	defer c.mu.Unlock()
	out := make(map[string]*EndpointMetrics, len(c.endpoints))
	for k, v := range c.endpoints {
		clone := *v
		clone.StatusCodes = make(map[int]int, len(v.StatusCodes))
		for sc, cnt := range v.StatusCodes {
			clone.StatusCodes[sc] = cnt
		}
		clone.CanaryHits = make(map[string]int, len(v.CanaryHits))
		for cv, cnt := range v.CanaryHits {
			clone.CanaryHits[cv] = cnt
		}
		latCopy := make([]float64, len(v.Latency.samples))
		copy(latCopy, v.Latency.samples)
		clone.Latency = LatencyStats{samples: latCopy}
		out[k] = &clone
	}
	return out
}

// GlobalErrorRate returns the aggregate error rate across all endpoints.
func (c *Collector) GlobalErrorRate() float64 {
	snap := c.Snapshot()
	total, errors := 0, 0
	for _, m := range snap {
		total += m.Total
		errors += m.ClientError + m.ServerError + m.Timeout
	}
	if total == 0 {
		return 0
	}
	return float64(errors) / float64(total)
}

// GlobalP99 returns the 99th percentile latency across all endpoints.
func (c *Collector) GlobalP99() float64 {
	snap := c.Snapshot()
	var all []float64
	for _, m := range snap {
		all = append(all, m.Latency.samples...)
	}
	if len(all) == 0 {
		return 0
	}
	sort.Float64s(all)
	idx := int(float64(len(all)) * 0.99)
	if idx >= len(all) {
		idx = len(all) - 1
	}
	return all[idx]
}
